File: 2487_remove_nodes_from_linked_list/removeNodes.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseLinkedList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        tail = None

        while head:
            currentNext = head.next
            head.next = tail
            tail = head
            head = currentNext
        return tail

    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:

        current = tail = self.reverseLinkedList(head)
        currentMax = current.val

        while current.next:
            if current.next.val < currentMax:
                current.next = current.next.next
            else:
                current = current.next
                currentMax = current.val

        return self.reverseLinkedList(tail)

File: 2487_remove_nodes_from_linked_list/removeNodes.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode head) {
        ListNode tail = null;

        while (head != null) {
            ListNode next = head.next;
            head.next = tail;
            tail = head;
            head = next;
        }

        return tail;
    }
    public ListNode removeNodes(ListNode head) {
        ListNode current = reverse(head);
        ListNode tail = current;
        int max = current.val;

        while (current.next != null ){
            if (current.next.val < max) {
                current.next = current.next.next;
            } else {
                current = current.next;
                max = current.val;
            }
        }

        return reverse(tail);
    }
}